### PR TITLE
Enable use of pollset_set for a channel call

### PR DESCRIPTION
**Pull request by dgquintas**

### CODE REVIEW

**Comment by dgquintas** [comment]
How can  `client_auth_filter.c:187] assertion failed: calld->pollset_set` fail (only) on Windows then `calld->pollset_set` is clearly initialized in `init_call_elem`? Is there anything different when it comes to windows going on (other than the compiler).

**Comment by dgquintas** [comment]
Oh, here we go... https://github.com/grpc/grpc/blob/master/src/core/lib/iomgr/pollset_set_windows.c#L40 
wut?

File: src/core/lib/channel/channel_stack.c [hunk 1]
```diff
@@ -189,29 +189,32 @@ void grpc_call_stack_init(grpc_exec_ctx *exec_ctx,
...
-                                 grpc_pollset *pollset) {
+void grpc_call_stack_set_pollset_or_pollset_set(
+    grpc_exec_ctx *exec_ctx, grpc_call_stack *call_stack, grpc_pollset *pollset,
+    grpc_pollset_set *or_pollset_set) {
   size_t count = call_stack->count;
   grpc_call_element *call_elems;
   char *user_data;
   size_t i;
 
+  GPR_ASSERT(!(pollset != NULL && or_pollset_set != NULL));
```
Comment (ctiller) [inline src/core/lib/channel/channel_stack.c]: makes my brain bleed: rename parameter? (I know I suggested the name)

Comment (ctiller) [inline src/core/lib/channel/channel_stack.c]: I've always kind of liked:

GPR_ASSERT((a == NULL) + (b == NULL) == 1);

Comment (dgquintas) [inline src/core/lib/channel/channel_stack.c]: Done and done. The new name isn't brilliant either, but...

File: src/core/ext/client_config/subchannel_call_holder.c [hunk 2]
```diff
@@ -68,6 +68,7 @@ void grpc_subchannel_call_holder_init(
   holder->waiting_ops_capacity = 0;
   holder->creation_phase = GRPC_SUBCHANNEL_CALL_HOLDER_NOT_CREATING;
   holder->owning_call = owning_call;
+  holder->pollset_set = grpc_pollset_set_create();
```
Comment (ctiller) [inline src/core/ext/client_config/subchannel_call_holder.c]: This is going to be quite bad for performance. The typical case is that we just want to set a pollset and not go ahead and create a new big container of object. Suggest bifurcating here and only creating the pollset_set when necessary.

Comment (dgquintas) [inline src/core/ext/client_config/subchannel_call_holder.c]: Done.

**Comment by ctiller** [comment]
Needs a merge.

Looking at the code now I'm a bit worried about pops too: the variable name is ambiguous as to whether it represents a grpc_pops, or a pointer to some operations. 

grpc_polling_entity?

Apologies for the bikeshed.

**Comment by dgquintas** [comment]
No worries :) I'm getting really good at grep+xargs+sed

Done. PTAL.

File: src/core/lib/iomgr/polling_entity.h [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+typedef struct grpc_polling_entity {
+  union {
+    grpc_pollset *pollset;
+    grpc_pollset_set *pollset_set;
+  } pollent;
+  enum pops_tag { POPS_NONE, POPS_POLLSET, POPS_POLLSET_SET } tag;
+} grpc_polling_entity;
+
+grpc_polling_entity grpc_pops_create_from_pollset_set(
```
Comment (ctiller) [inline src/core/lib/iomgr/polling_entity.h]: We probably want to rename these functions for consistency also

Comment (dgquintas) [inline src/core/lib/iomgr/polling_entity.h]: Woops. Done.

**Comment by ctiller** [comment]
LGTM... feel free to merge on green.

**Comment by jtattermusch** [comment]
Are we 100% the ruby failure is unrelated? https://grpc-testing.appspot.com/job/gRPC_pull_requests/9612/config=opt,language=ruby,platform=linux/testReport/junit/(root)/tests/tools_run_tests_run_ruby_sh/

**Comment by dgquintas** [comment]
Re-running to make sure

**Comment by dgquintas** [comment]
FWIW, a local run under docker passes fine on my machine (yes, anecdata, let's see what jenkins thinks).

**Comment by dgquintas** [comment]
The current jenkins run (without any further changes to this PR) is passing for ruby (https://grpc-testing.appspot.com/job/gRPC_pull_requests/9633/). There are a couple red exclamation marks for python/windows due to infrastructure hiccups. I'd argue that this run plus the previous one show this PR is good to be merged.

File: src/core/lib/iomgr/timer.c [hunk 4]
```diff
@@ -278,7 +278,8 @@ static int refill_queue(shard_type *shard, gpr_timespec now) {
   return !grpc_timer_heap_is_empty(&shard->heap);
 }
 
-/* This pops the next non-cancelled timer with deadline <= now from the queue,
+/* This pollent the next non-cancelled timer with deadline <= now from the
```
Comment (ashoemaker) [inline src/core/lib/iomgr/timer.c]: Unintended?

Comment (dgquintas) [inline src/core/lib/iomgr/timer.c]: Ugh, yes. The perils of global sed search-and-replace. I'll send a fixing PR momentarily.

Comment (dgquintas) [inline src/core/lib/iomgr/timer.c]: Done! https://github.com/grpc/grpc/pull/6832 Thanks!

**Pull request merged**